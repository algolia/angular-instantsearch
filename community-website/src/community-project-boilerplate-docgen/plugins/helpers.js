const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
