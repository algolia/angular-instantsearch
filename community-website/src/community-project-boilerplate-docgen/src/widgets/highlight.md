---
title: Widgets - Highlight
layout: widget.pug
---

# Highlight

## Description

Renders any attribute from an hit into its highlighted form when relevant.

## Options

| Attributes  | Type     | Description
| -           | -        | -
| `attribute` | `string` | The location of the highlighted attribute in the hit
| `hit`       | `{}`     | The hit object containing the highlighted attribute
| `tagName?`  | `string` | The tag to be used for highlighted parts of the hit

## Code example

```html
<ng-ais-hits>
  <ng-template let-hits="hits">
    <div *ngFor="let hit of hits">
      Hit {{hit.objectID}}:
      <ng-ais-highlight attribute="name" [hit]="hit">
      </ng-ais-highlight>
    </div>
  </ng-template>
</ng-ais-hits>
```
