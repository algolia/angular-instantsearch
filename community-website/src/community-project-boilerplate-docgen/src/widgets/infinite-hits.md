---
title: Widgets - Infinite hits
layout: widget.pug
---

# Infinite Hits

## Description

Displays a list of hits.
You can use the directive `<ng-template></ng-template>` to customize the output.

## Options

| Attribures            | Type     | Description
| -                     | -        | -
| `showMoreLabel?`      | `string` | Label used on the show more button.

## Code example

```html
<ais-infinite-hits>
  <ng-template
    let-hits="hits"
    let-results="results"
  >
    <!-- No results message -->
    <div *ngIf="hits.length === 0">
      No results found matching <strong>{{results.query}}</strong>.
    </div>

    <!-- Hit template -->
    <div *ngFor="let hit of hits">
      <h1>
        <ais-highlight
          attribute="title"
          [hit]="hit"
        >
        </ais-highlight>
      </h1>

      <p>{{hit.description}}</p>
    </div>
  </ng-template>
</ais-infinite-hits>
```
