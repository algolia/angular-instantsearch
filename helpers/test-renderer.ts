import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
