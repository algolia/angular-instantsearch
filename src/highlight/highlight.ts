import { Component, Input } from "@angular/core";
import { isPlainObject, escape, get } from "lodash-es";
import { bem } from "../utils";

@Component({
  selector: "ais-highlight",
  template: `<span [class]="cx()" [innerHtml]="content"></span>`
})
export class NgAisHighlight {
  @Input() attribute: string;
  @Input() hit: { _highlightResult?: {}; label?: string; highlighted?: string };
  @Input() tagName: string = "em";

  cx = bem("Highlight");

  get content() {
    if (this.attribute === "highlighted") {
      return this.hit.highlighted
        ? this.replaceWithTagName(this.hit.highlighted)
        : this.hit.label;
    }

    if (this.hit.hasOwnProperty("_highlightResult")) {
      const attributeHighlighted = get(
        this.hit._highlightResult,
        this.attribute
      );

      // check that the attributeHighlighted is a string
      if (
        isPlainObject(attributeHighlighted) &&
        typeof attributeHighlighted.value === "string"
      ) {
        return this.replaceWithTagName(attributeHighlighted.value);
      }
    }

    const fallback = get(this.hit, this.attribute);
    if (!fallback) {
      console.warn(
        `Could not find attribute [${
          this.attribute
        }] into hit object, will display an empty string.`
      );

      return "";
    }

    return fallback;
  }

  replaceWithTagName(value: string) {
    return value
      .replace(
        new RegExp("<em>", "g"),
        `<${this.tagName} class="${this.cx("highlighted")}">`
      )
      .replace(new RegExp("</em>", "g"), `</${this.tagName}>`);
  }
}
