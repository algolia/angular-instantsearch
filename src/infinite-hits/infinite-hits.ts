import {
  Component,
  ContentChild,
  Input,
  TemplateRef,
  Inject,
  forwardRef,
} from '@angular/core';

import { connectInfiniteHits } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { noop } from '../utils';

@Component({
  selector: 'ais-infinite-hits',
  template: `
    <div [class]="cx()">
      <ng-container *ngTemplateOutlet="template; context: state"></ng-container>

      <!-- default rendering if no template specified -->
      <div *ngIf="!template">
        <ul [class]="cx('list')">
          <li
            [class]="cx('item')"
            *ngFor="let hit of state.hits"
          >
            <ais-highlight attribute="name" [hit]="hit">
            </ais-highlight>
          </li>
        </ul>
      </div>

      <button
        [ngClass]="[cx('loadMore'), this.state.isLastPage ? cx('loadMore', 'disabled') : '']"
        (click)="showMore($event)"
        [disabled]="state.isLastPage"
        *ngIf="!template"
      >
        {{showMoreLabel}}
      </button>
    </div>
  `,
})
export class NgAisInfiniteHits extends BaseWidget {
  @ContentChild(TemplateRef) public template?: any;

  // render options
  @Input() public showMoreLabel: string = 'Show more results';
  @Input() public transformItems?: Function;

  // inner widget state returned from connector
  public state: {
    hits: {}[];
    isLastPage: boolean;
    showMore: Function;
    results: {};
  } = {
    hits: [],
    isLastPage: false,
    showMore: noop,
    results: {},
  };

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('InfiniteHits');
    this.createWidget(connectInfiniteHits, { escapeHits: true });
  }

  public showMore(event: MouseEvent) {
    event.preventDefault();
    this.state.showMore();
  }

  updateState = (state, isFirstRendering: boolean) => {
    if (isFirstRendering) return;

    this.state = {
      ...state,
      results: state.results,
      hits:
        typeof this.transformItems === 'function'
          ? this.transformItems(state.hits)
          : state.hits,
    };
  };
}
