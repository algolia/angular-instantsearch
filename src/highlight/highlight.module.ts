import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisHighlight } from './highlight';

@NgModule({
  declarations: [NgAisHighlight],
  entryComponents: [NgAisHighlight],
  exports: [NgAisHighlight],
  imports: [CommonModule],
})
export class NgAisHighlightModule {}
