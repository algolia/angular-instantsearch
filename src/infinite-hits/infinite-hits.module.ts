import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisHighlightModule } from '../highlight/highlight.module';
import { NgAisInfiniteHits } from './infinite-hits';

@NgModule({
  declarations: [NgAisInfiniteHits],
  entryComponents: [NgAisInfiniteHits],
  exports: [NgAisInfiniteHits],
  imports: [CommonModule, NgAisHighlightModule],
})
export class NgAisInfiniteHitsModule {}
