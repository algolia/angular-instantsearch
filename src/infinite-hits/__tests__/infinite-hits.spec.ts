import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisInfiniteHits } from "../infinite-hits";
import { NgAisHighlight } from "../../highlight/highlight";

const defaultState = {
  hits: [
    { name: "foo", description: "foo" },
    { name: "bar", description: "bar" },
    { name: "foobar", description: "foobar" },
    { name: "barfoo", description: "barfoo" }
  ],
  showMore: jest.fn(),
  isLastPage: false
};

const render = createRenderer({
  defaultState,
  template: "<ais-infinite-hits></ais-infinite-hits>",
  TestedWidget: NgAisInfiniteHits,
  additionalDeclarations: [NgAisHighlight]
});

describe("InfiniteHits", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should call `showMore()` on button click", () => {
    const showMore = jest.fn();
    const fixture = render({ showMore });

    const button = fixture.debugElement.nativeElement.querySelector("button");
    button.click();

    expect(button.disabled).toBeFalsy();
    expect(showMore).toHaveBeenCalled();
  });

  it("should disable `showMore` button", () => {
    const showMore = jest.fn();
    const fixture = render({ showMore, isLastPage: true });

    const button = fixture.debugElement.nativeElement.querySelector("button");
    button.click();

    expect(button.disabled).toBeTruthy();
    expect(showMore).not.toHaveBeenCalled();
  });
});
