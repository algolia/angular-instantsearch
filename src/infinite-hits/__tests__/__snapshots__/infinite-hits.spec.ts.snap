// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteHits renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgAisInfiniteHits]}
>
  
        <ais-instantsearch>
    
          
    <ais-infinite-hits>
      
      
      <div
        class="ais-InfiniteHits"
      >
        
        
        
        
  
        
        
        
        
        <div>
          
          
          <ul
            class="ais-InfiniteHits-list"
          >
            
            
            
            <li
              class="ais-InfiniteHits-item"
            >
              
              
              <ais-highlight
                attribute="name"
                ng-reflect-attribute="name"
                ng-reflect-hit="[object Object]"
              >
                <span
                  class="ais-Highlight"
                >
                  foo
                </span>
              </ais-highlight>
              
            
            </li>
            <li
              class="ais-InfiniteHits-item"
            >
              
              
              <ais-highlight
                attribute="name"
                ng-reflect-attribute="name"
                ng-reflect-hit="[object Object]"
              >
                <span
                  class="ais-Highlight"
                >
                  bar
                </span>
              </ais-highlight>
              
            
            </li>
            <li
              class="ais-InfiniteHits-item"
            >
              
              
              <ais-highlight
                attribute="name"
                ng-reflect-attribute="name"
                ng-reflect-hit="[object Object]"
              >
                <span
                  class="ais-Highlight"
                >
                  foobar
                </span>
              </ais-highlight>
              
            
            </li>
            <li
              class="ais-InfiniteHits-item"
            >
              
              
              <ais-highlight
                attribute="name"
                ng-reflect-attribute="name"
                ng-reflect-hit="[object Object]"
              >
                <span
                  class="ais-Highlight"
                >
                  barfoo
                </span>
              </ais-highlight>
              
            
            </li>
            
          
          </ul>
          
        
        </div>
        
  
        
        
        <button
          class="ais-InfiniteHits-showMore"
        >
          
          Show more results
        
        </button>
        
      
      </div>
      
    
    </ais-infinite-hits>
    
        
  </ais-instantsearch>
      
</ng-component>
`;

exports[`InfiniteHits renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisInfiniteHits]}
>
  
        <ais-instantsearch>
    
          
    <ais-infinite-hits>
      
      
      <div
        class="ais-InfiniteHits"
      >
        
        
        
        
  
        
        
        
        
        <div>
          
          
          <ul
            class="ais-InfiniteHits-list"
          >
            
            
            
            
          
          </ul>
          
        
        </div>
        
  
        
        
        <button
          class="ais-InfiniteHits-showMore"
        >
          
          Show more results
        
        </button>
        
      
      </div>
      
    
    </ais-infinite-hits>
    
        
  </ais-instantsearch>
      
</ng-component>
`;
